--- initramfs-init
+++ initramfs-init-new
@@ -196,10 +196,16 @@
 			echo "ERROR: DHCP requested but not present in initrd"
 			return 1
 		fi
+
+		ebegin "Set hostname based on hardware addr. ($device)"
+		HOSTNAME=$(tr -d ':' < /sys/class/net/$device/address)
+		hostname "$HOSTNAME"
+		eend $?
+
 		# automatic configuration
 		ebegin "Obtaining IP via DHCP ($device)..."
 		ifconfig $device 0.0.0.0
-		udhcpc -i $device -f -q
+		udhcpc -i $device -f -q -F $HOSTNAME
 		eend $?
 	else
 		# manual configuration
@@ -566,6 +572,8 @@
 mount -o bind /sys $sysroot/sys
 mount -o bind /proc $sysroot/proc
 mount -o bind /dev $sysroot/dev
+# fix error CAP_SYS_ADMIN
+sysctl -w kernel.grsecurity.chroot_caps=0
 if [ -n "$ovlfiles" ]; then
 	apk add --root $sysroot $repo_opt $apkflags $pkgs <$ovlfiles
 else
@@ -608,6 +616,11 @@
 
 ! [ -f "$sysroot"/etc/resolv.conf ] && [ -f /etc/resolv.conf ] && \
   cp /etc/resolv.conf "$sysroot"/etc
+
+# setup hostname
+if [ -n "$HOSTNAME" ]; then
+	echo "$HOSTNAME" >$sysroot/etc/hostname
+fi
 
 # setup bootchart for switch_root
 chart_init=""
